### PR TITLE
plugin: use pkg/errors in more places

### PR DESCRIPTION
**Pull request by tiborvass**
Also provide stack trace output in daemon logs.

Signed-off-by: Tibor Vass <tibor@docker.com>

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]
LGTM (if green)

**Comment by vieux** [comment]
LGTM

File: plugin/manager.go [hunk 1]
```diff
@@ -237,7 +237,7 @@ func (pm *Manager) save(p *v2.Plugin) error {
 		return errors.Wrap(err, "failed to marshal plugin json")
 	}
 	if err := ioutils.AtomicWriteFile(filepath.Join(pm.config.Root, p.GetID(), configFileName), pluginJSON, 0600); err != nil {
-		return err
+		return errors.Wrap(err, "failed to write atomically plugin json")
```
Comment (anusha-ragunathan) [inline plugin/manager.go]: Should be "failed to atomically write plugin json"

**Comment by anusha-ragunathan** [comment]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Comment by aaronlehmann** [comment]
After this change I'm seeing stack traces being dumped to the logs when a swarm command is used while not in swarm mode:

```
ERRO[0064] Handler for POST /v1.26/services/create returned error: This node is not a swarm manager. Use "docker swarm init" or "docker swarm join" to connect this node to swarm and try again.
github.com/docker/docker/daemon/cluster.(*Cluster).errNoManager
	/go/src/github.com/docker/docker/daemon/cluster/cluster.go:790
github.com/docker/docker/daemon/cluster.(*Cluster).CreateService
	/go/src/github.com/docker/docker/daemon/cluster/cluster.go:885
github.com/docker/docker/api/server/router/swarm.(*swarmRouter).createService
	/go/src/github.com/docker/docker/api/server/router/swarm/cluster_routes.go:172
github.com/docker/docker/api/server/router/swarm.(*swarmRouter).(github.com/docker/docker/api/server/router/swarm.createService)-fm
	/go/src/github.com/docker/docker/api/server/router/swarm/cluster.go:36
github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1
	/go/src/github.com/docker/docker/api/server/middleware/experimental.go:27
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1
	/go/src/github.com/docker/docker/api/server/middleware/version.go:47
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1
	/go/src/github.com/docker/docker/pkg/authorization/middleware.go:43
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1
	/go/src/github.com/docker/docker/api/server/middleware/debug.go:53
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1
	/go/src/github.com/docker/docker/api/server/server.go:140
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:1726
github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
	/go/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:103
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP
	/go/src/github.com/docker/docker/api/server/router_swapper.go:29
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2202
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1579
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2086
```

I'm strongly -1 on the logging change. I think we should only dump stack traces to logs when they are necessary for debugging. They are confusing to users and make it look like the daemon has crashed.

**Comment by aaronlehmann** [comment]
It also happens with `docker inspect`. The CLI tries a few different endpoints. The `images` endpoint dumps a stack trace when it fails:

```
ERRO[0010] Handler for GET /v1.26/images/foo/json returned error: no such id: foo
no such image: foo
github.com/docker/docker/daemon.(*Daemon).LookupImage
	/go/src/github.com/docker/docker/daemon/image_inspect.go:17
github.com/docker/docker/api/server/router/image.(*imageRouter).getImagesByName
	/go/src/github.com/docker/docker/api/server/router/image/image_routes.go:238
github.com/docker/docker/api/server/router/image.(*imageRouter).(github.com/docker/docker/api/server/router/image.getImagesByName)-fm
	/go/src/github.com/docker/docker/api/server/router/image/image.go:39
github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1
	/go/src/github.com/docker/docker/api/server/middleware/experimental.go:27
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1
	/go/src/github.com/docker/docker/api/server/middleware/version.go:47
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1
	/go/src/github.com/docker/docker/pkg/authorization/middleware.go:43
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1
	/go/src/github.com/docker/docker/api/server/middleware/debug.go:22
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1
	/go/src/github.com/docker/docker/api/server/server.go:140
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:1726
github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
	/go/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:103
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP
	/go/src/github.com/docker/docker/api/server/router_swapper.go:29
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2202
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1579
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2086
```

**Comment by anusha-ragunathan** [comment]
I have to agree with aaron here. I tried `docker plugin upgrade` with a non-existent image and get a daemon stacktrace, which is quite alarming.

```
ERRO[0055] Handler for GET /v1.26/plugins/foo/bar/json returned error: plugin "foo/bar" not found
github.com/docker/docker/plugin.(*Store).resolvePluginID
        /go/src/github.com/docker/docker/plugin/store.go:260
github.com/docker/docker/plugin.(*Store).GetV2Plugin
        /go/src/github.com/docker/docker/plugin/store.go:44
github.com/docker/docker/plugin.(*Manager).Inspect
        /go/src/github.com/docker/docker/plugin/backend_linux.go:82
github.com/docker/docker/api/server/router/plugin.(*pluginRouter).inspectPlugin
        /go/src/github.com/docker/docker/api/server/router/plugin/plugin_routes.go:318
github.com/docker/docker/api/server/router/plugin.(*pluginRouter).(github.com/docker/docker/api/server/router/plugin.inspectPlugin)-fm
        /go/src/github.com/docker/docker/api/server/router/plugin/plugin.go:28
github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1
        /go/src/github.com/docker/docker/api/server/middleware/experimental.go:27
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1
        /go/src/github.com/docker/docker/api/server/middleware/version.go:47
github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1
        /go/src/github.com/docker/docker/pkg/authorization/middleware.go:43
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1
        /go/src/github.com/docker/docker/api/server/server.go:140
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:1726
github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
        /go/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:103
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP
        /go/src/github.com/docker/docker/api/server/router_swapper.go:29
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2202
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1579
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:2086
```

**Comment by thaJeztah** [comment]
can we enable this behavior only if debug is enabled? Or should we change it back altogether? I agree these logs look alarming if you're not aware of this change

**Comment by tiborvass** [comment]
@aaronlehmann @anusha-ragunathan I understand the concern and didn't mean to make it this way. However, I do think it is useful to have the stacktrace for debugging obscure errors. Would you be okay to do what @thaJeztah suggested, that is have this behavior only in debug mode?

**Comment by aaronlehmann** [comment]
Even in debug mode, it seems like a lot of noise. For every obscure error there's going to be hundreds of "normal" cases where the handler returned a HTTP-level error.

Would it be possible to discriminate based on HTTP error code - i.e. only show the stack trace for 500 errors and not routine error codes like 404?

**Pull request merged**